File: backend/app/api/catalogs.py
"""Catalogue endpoints (RF-034, M04).

Three audiences, and they ask different questions:

* **A device** asks for the delta: everything above the revision it holds, retired values included.
* **A screen** asks for the resolved list: the active values this unit can choose from.
* **An administrator** writes. Only the ones the platform owns: a catalogue the ERP maintains
  refuses a hand edit, because a value typed over one the connector will overwrite tonight
  disappears without explanation.
"""

from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.auth.dependencies import current_principal, require_roles, unit_scope
from app.auth.principal import Role
from app.catalogs import service as catalogs
from app.catalogs.service import MAX_DELTA
from app.infra.database import get_session
from app.org.models import BusinessUnit
from app.org.service import UnknownBusinessUnitError, get_business_unit_by_code

# El segmento literal `catalog/` antes del código no es decoración: sin él,
# `/units/GYE/delta` casa con `/units/{unit_code}/{catalog_code}` y el delta del dispositivo
# resuelve al catálogo inexistente «delta». Lo encontró un test, y la forma de que no vuelva a
# pasar no es ordenar las rutas —deja la ambigüedad latente para un catálogo que se llame así—
# sino que ningún código de catálogo pueda ocupar la posición de un verbo.
router = APIRouter(
    prefix="/api/v1/catalogs", tags=["catalogs"], dependencies=[Depends(current_principal)]
)

SessionDep = Annotated[Session, Depends(get_session)]

EDITORS = (Role.FUNCTIONAL_ADMIN, Role.IT_ADMIN)


def _unit(session: Session, code: str) -> BusinessUnit:
    try:
        return get_business_unit_by_code(session, code)
    except UnknownBusinessUnitError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc


class EntryIn(BaseModel):
    """One value. No author: it is the token's subject."""

    code: str = Field(min_length=1, max_length=64)
    label: str = Field(min_length=1, max_length=255)
    synonyms: list[str] = Field(default_factory=list, max_length=50)
    parent_code: str | None = Field(default=None, max_length=64)
    attributes: dict[str, Any] = Field(default_factory=dict)
    sort_order: int = 0
    active: bool = True


@router.get("/", summary="Los catálogos que la plataforma sirve, con sus versiones (RF-034)")
def index(session: SessionDep) -> dict[str, Any]:
    """One number per catalogue, which is what a device compares before asking for a delta."""
    counts = catalogs.entry_counts(session)
    rows = []
    for catalog in catalogs.list_catalogs(session):
        live, gone = counts.get(catalog.code, (0, 0))
        rows.append(
            {
                "code": catalog.code,
                "title": catalog.title,
                "source": catalog.source,
                "version": catalog.version,
                "note": catalog.note,
                "active_entries": live,
                "retired_entries": gone,
                #: An empty catalogue is legitimate and has to explain itself: the political
                #: division is deliberately empty until the official list is loaded.
                "empty": live == 0,
            }
        )
    return {"catalogs": rows, "gis_backed": list(catalogs.GIS_BACKED)}


@router.get(
    "/units/{unit_code}/catalog/{catalog_code}",
    dependencies=[Depends(unit_scope)],
    summary="Los valores que puede elegir un técnico de esta unidad (RF-034)",
)
def resolved(session: SessionDep, unit_code: str, catalog_code: str) -> dict[str, Any]:
    """The national list with this unit's own additions folded in; a local code overrides."""
    unit = _unit(session, unit_code)
    try:
        return catalogs.resolve(session, catalog_code, unit).as_dict()
    except catalogs.UnknownCatalogError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc


@router.get(
    "/units/{unit_code}/delta",
    dependencies=[Depends(unit_scope)],
    summary="Lo que cambió desde una revisión, lápidas incluidas (RF-034)",
)
def delta(
    session: SessionDep,
    unit_code: str,
    since: Annotated[int, Query(ge=0)] = 0,
    codes: Annotated[list[str] | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=MAX_DELTA)] = MAX_DELTA,
) -> dict[str, Any]:
    """The retired values travel too: a delta that skipped them would leave a phone offering a code
    the utility withdrew, and nobody would notice because the value looks ordinary."""
    unit = _unit(session, unit_code)
    return catalogs.delta(session, since=since, unit=unit, codes=codes, limit=limit).as_dict()


@router.put(
    "/catalog/{catalog_code}/entries/{entry_code}",
    dependencies=[Depends(require_roles(*EDITORS))],
    summary="Crear o corregir un valor nacional (RF-034)",
)
def upsert_national(
    session: SessionDep,
    catalog_code: str,
    entry_code: str,
    payload: EntryIn,
    principal: Annotated[Any, Depends(current_principal)] = None,
) -> dict[str, Any]:
    if payload.code != entry_code:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            f"el código de la ruta («{entry_code}») y el del cuerpo («{payload.code}») difieren",
        )
    try:
        row = catalogs.upsert_entry(
            session,
            catalog_code,
            entry_code=entry_code,
            label=payload.label,
            synonyms=payload.synonyms,
            parent_code=payload.parent_code,
            attributes=payload.attributes,
            sort_order=payload.sort_order,
            active=payload.active,
            actor=principal.subject,
        )
    except catalogs.NotEditableError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc)) from exc
    except catalogs.UnknownCatalogError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    session.commit()
    return {"catalog": catalog_code, "code": row.code, "revision": row.revision}


@router.put(
    "/units/{unit_code}/catalog/{catalog_code}/entries/{entry_code}",
    dependencies=[Depends(unit_scope), Depends(require_roles(*EDITORS))],
    summary="Crear o corregir un valor propio de una unidad (RF-034)",
)
def upsert_local(
    session: SessionDep,
    unit_code: str,
    catalog_code: str,
    entry_code: str,
    payload: EntryIn,
    principal: Annotated[Any, Depends(current_principal)] = None,
) -> dict[str, Any]:
    """A unit's value with a national one's code overrides it rather than duplicating it."""
    unit = _unit(session, unit_code)
    try:
        row = catalogs.upsert_entry(
            session,
            catalog_code,
            entry_code=entry_code,
            label=payload.label,
            unit=unit,
            synonyms=payload.synonyms,
            parent_code=payload.parent_code,
            attributes=payload.attributes,
            sort_order=payload.sort_order,
            active=payload.active,
            actor=principal.subject,
        )
    except catalogs.NotEditableError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc)) from exc
    except catalogs.UnknownCatalogError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    session.commit()
    return {"catalog": catalog_code, "code": row.code, "revision": row.revision}


@router.delete(
    "/catalog/{catalog_code}/entries/{entry_code}",
    dependencies=[Depends(require_roles(*EDITORS))],
    summary="Retirar un valor nacional (RF-034)",
)
def retire_national(
    session: SessionDep,
    catalog_code: str,
    entry_code: str,
    principal: Annotated[Any, Depends(current_principal)] = None,
) -> dict[str, Any]:
    """Deactivates; there is no delete. The tombstone has to reach the device."""
    try:
        row = catalogs.retire_entry(session, catalog_code, entry_code, actor=principal.subject)
    except catalogs.UnknownCatalogError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    session.commit()
    return {"catalog": catalog_code, "code": row.code, "active": row.active}

File: web/src/api/catalogs.ts
/**
 * Catalogue client (RF-034).
 *
 * The device's delta endpoint is not used from here — a browser has nothing to sync — so this
 * covers the two things a screen needs: the index with one version per catalogue, and the resolved
 * list for one business unit.
 */

import { ApiError } from './planning';
import { authHeaders, notifyExpired } from './session';

export interface CatalogSummary {
  code: string;
  title: string;
  /** 'manual' or 'integracion': who owns the values. */
  source: string;
  version: number;
  /** Why it is empty, when it is. The server's words. */
  note: string | null;
  active_entries: number;
  retired_entries: number;
  empty: boolean;
}

export interface CatalogIndex {
  catalogs: CatalogSummary[];
  /** Codes served by the unit's synced GIS metadata, not by this table (RF-304). */
  gis_backed: string[];
}

export interface CatalogEntryView {
  code: string;
  label: string;
  synonyms: string[];
  parent_code: string | null;
  attributes: Record<string, unknown>;
  /** True when the value is this unit's rather than the national list's. */
  local: boolean;
}

export interface ResolvedCatalog {
  code: string;
  title: string;
  source: string;
  version: number;
  note: string | null;
  empty: boolean;
  entries: CatalogEntryView[];
}

const BASE = '/api/v1/catalogs';

async function call<T>(path: string, init: RequestInit = {}): Promise<T> {
  const response = await fetch(path, {
    ...init,
    headers: {
      ...authHeaders(),
      ...(init.body ? { 'Content-Type': 'application/json' } : {}),
      ...(init.headers ?? {}),
    },
  });
  if (!response.ok) {
    if (response.status === 401) notifyExpired();
    let detail = `${response.status} ${response.statusText}`;
    try {
      const body = (await response.json()) as { detail?: unknown };
      if (typeof body.detail === 'string') detail = body.detail;
    } catch {
      // No JSON body; the status line stands.
    }
    throw new ApiError(response.status, detail);
  }
  return (await response.json()) as T;
}

export function fetchIndex(signal?: AbortSignal): Promise<CatalogIndex> {
  return call<CatalogIndex>(`${BASE}/`, { signal });
}

export function fetchResolved(
  businessUnit: string,
  code: string,
  signal?: AbortSignal,
): Promise<ResolvedCatalog> {
  return call<ResolvedCatalog>(
    `${BASE}/units/${encodeURIComponent(businessUnit)}/catalog/${encodeURIComponent(code)}`,
    { signal },
  );
}

export interface EntryPayload {
  code: string;
  label: string;
  synonyms?: string[];
  parent_code?: string | null;
  attributes?: Record<string, unknown>;
  sort_order?: number;
  active?: boolean;
}

/** Write a national value. The author is the token's subject. */
export function saveNationalEntry(code: string, entry: EntryPayload): Promise<unknown> {
  return call(
    `${BASE}/catalog/${encodeURIComponent(code)}/entries/${encodeURIComponent(entry.code)}`,
    { method: 'PUT', body: JSON.stringify(entry) },
  );
}

/** Write a value only this unit sees. A national code it repeats is overridden, not duplicated. */
export function saveLocalEntry(
  businessUnit: string,
  code: string,
  entry: EntryPayload,
): Promise<unknown> {
  return call(
    `${BASE}/units/${encodeURIComponent(businessUnit)}/catalog/${encodeURIComponent(code)}/entries/${encodeURIComponent(entry.code)}`,
    { method: 'PUT', body: JSON.stringify(entry) },
  );
}

/** Deactivates; there is no delete. The tombstone has to reach the device. */
export function retireEntry(code: string, entryCode: string): Promise<unknown> {
  return call(
    `${BASE}/catalog/${encodeURIComponent(code)}/entries/${encodeURIComponent(entryCode)}`,
    { method: 'DELETE' },
  );
}
